### PR TITLE
Fix CA2237: Mark ISerializable types with SerializableAttribute

### PR DESCRIPTION
**Pull request by heng-liu**
## Bug

Fixes:https://github.com/NuGet/Client.Engineering/issues/651

Regression? Last working version:

## Description
1. Add `SerializableAttribute ` attribute for all exception class reported with CA2237 violation.
2. Remove the suppression entry of CA2237.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [x] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreSpecException.cs [hunk 1]
```diff
@@ -12,6 +12,7 @@ namespace NuGet.Commands
     /// <summary>
     /// DG v2 related validation error.
     /// </summary>
+    [SerializableAttribute]
     public class RestoreSpecException : Exception
     {
         public IEnumerable<string> Files { get; }
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreSpecException.cs:18]: This class needs to implement `ISerializable` properly for `Files` to round trip.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommandException.cs [hunk 2]
```diff
@@ -9,6 +9,7 @@ namespace NuGet.Commands
     /// <summary>
     /// Holds an <see cref="IRestoreLogMessage"/> and returns the message for the exception.
     /// </summary>
+    [SerializableAttribute]
     public class RestoreCommandException : Exception, ILogMessageException
     {
         private readonly IRestoreLogMessage _logMessage;
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommandException.cs:15]: This class needs to implement `ISeriablizable` properly for `_logMessage` to round trip.

File: src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandException.cs [hunk 3]
```diff
@@ -9,6 +9,7 @@ namespace NuGet.Commands
     /// <summary>
     /// Holds an <see cref="ILogMessage"/> and returns the message for the exception.
     /// </summary>
+    [SerializableAttribute]
     public sealed class SignCommandException : Exception, ILogMessageException
     {
         private readonly ILogMessage _logMessage;
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/SignCommand/SignCommandException.cs:15]: This class needs to implement `ISeriablizable` properly for `_logMessage` to round-trip

File: src/NuGet.Core/NuGet.Common/CommandLineArgumentCombinationException.cs [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
 namespace NuGet.Common
 {
+    [SerializableAttribute]
     public class CommandLineArgumentCombinationException : Exception, ILogMessageException
     {
         private readonly ILogMessage _logMessage;
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Common/CommandLineArgumentCombinationException.cs:11]: This class needs to implement `ISeriablizable` properly for `_logMessage` to round trip.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/PackageReferenceRollbackException.cs [hunk 5]
```diff
@@ -13,6 +13,7 @@ namespace NuGet.PackageManagement
     /// and additional log messages with specific information on
     /// what caused the rollback.
     /// </summary>
+    [SerializableAttribute]
     public class PackageReferenceRollbackException : InvalidOperationException
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/PackageReferenceRollbackException.cs:17]: This class needs to implement `ISerializable` properly for `LogMessages` to round trip.

File: src/NuGet.Core/NuGet.Packaging/Core/PackagingException.cs [hunk 6]
```diff
@@ -6,6 +6,7 @@
 
 namespace NuGet.Packaging.Core
 {
+    [SerializableAttribute]
     public class PackagingException : Exception, ILogMessageException
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Core/PackagingException.cs:10]: This class needs to implement `ISerializable` properly for `_logMessage` to round trip

File: src/NuGet.Core/NuGet.Packaging/Signing/SignatureException.cs [hunk 7]
```diff
@@ -10,6 +10,7 @@
 
 namespace NuGet.Packaging.Signing
 {
+    [SerializableAttribute]
     public class SignatureException : PackagingException
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Packaging/Signing/SignatureException.cs:14]: This class needs to implement `ISerializable` properly for all the properties to round trip.

File: src/NuGet.Core/NuGet.ProjectModel/FileFormatException.cs [hunk 8]
```diff
@@ -8,6 +8,7 @@
 
 namespace NuGet.ProjectModel
 {
+    [SerializableAttribute]
     public class FileFormatException : Exception
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.ProjectModel/FileFormatException.cs:12]: This class needs to implement `ISerializable` properly for the properties to round-trip.

File: src/NuGet.Core/NuGet.Protocol/Exceptions/PackageNotFoundProtocolException.cs [hunk 9]
```diff
@@ -11,6 +11,7 @@ namespace NuGet.Protocol
     /// <summary>
     /// Thrown when a package cannot be found on a feed.
     /// </summary>
+    [SerializableAttribute]
     public class PackageNotFoundProtocolException : InvalidCacheProtocolException
```
Comment (zivkan) [inline src/NuGet.Core/NuGet.Protocol/Exceptions/PackageNotFoundProtocolException.cs:15]: This class needs to implement `ISerializable` properly for `PackageIdentity` to round trip.

**Review by zivkan — COMMENTED** [review]
Two things.

Firstly, by convention we usually leave the word `Attribute` off the name when using attributes. So, it's better to use `[Serializable]`, rather than `[SerializableAttribute]`.

Secondly, `SerializableAttribute` means that the class implements the interface `ISerializable`, which `System.Exception` does, and hence all of our custom exception types inherit it. The [docs on `ISerializable`, particularly the remarks section](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.serialization.iserializable?view=net-5.0#remarks) explain what it means to implement correctly. Basicly, we need to override the method `GetObjectData`, including calling the base type's implementation, where we write all our custom exception's member fields and properties. We also need a constructor with args `(SerializationInfo info, StreamingContext context)` where we read those fields and properties that we wrote in the other method.

I can't remember if it's necessary in exceptions that do not have any fields/properties. I guess we still need the constructor so that the deserializer can create an instance of our type, but I can't remember. It's been a long time since I used `System.Runtime.Serialization`.

**Comment by heng-liu** [comment]
[Testing bot]: those comments will be addressed this sprint.

**Pull request closed without merging**